pkg/volume/util.go

```diff
@@ -148,3 +149,66 @@ func CalculateTimeoutForVolume(minimumTimeout, timeoutIncrement int, pv *api.Per
 func RoundUpSize(volumeSizeBytes int64, allocationUnitBytes int64) int64 {
 	return (volumeSizeBytes + allocationUnitBytes - 1) / allocationUnitBytes
 }
+
+// VolumeGuard may be (optionally) used by volume plugins to synchronize
+// SetUp and TearDown operation on the same volume. Most plugins are not
+// thread safe and bad things can happend, when they SetUp and TearDown
+// the same volume at the same time. With a VolumeGuard, they are thread safe.
+type VolumeGuard struct {
```

Comment:
This is very similar to [`pkg/util/keymutex/keymutex.go` ](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/gce_util.go) which is used for the same purpose for GCE PDs. It would make sense to reuse that code and modify it if needed.